caches/clan.go

```diff
@@ -0,0 +1,105 @@
+package caches
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	gocache "github.com/patrickmn/go-cache"
+	"github.com/topfreegames/khan/models"
+)
+
+// ClansSummaries represents a cache for the RetrieveClansSummaries operation.
+type ClansSummaries struct {
+	// Cache points to an instance of gocache.Cache used as the backend cache object.
+	Cache *gocache.Cache
+
+	// The cache for one clan will live by a random amount of time within [TTL - TTLRandomError, TTL + TTLRandomError].
+	TTL            time.Duration
+	TTLRandomError time.Duration
+}
+
+// GetClansSummaries is a cache in front of models.GetClansSummaries() with the exact same interface.
+// The map[string]interface{} return type represents a summary of one clan with the following keys/values:
+// "membershipCount":  int
+// "publicID":         string
+// "metadata":         map[string]interface{} (user-defined arbitrary JSON object with clan metadata)
+// "name":             string
+// "allowApplication": bool
+// "autoJoin":         bool
+// TODO: replace this map with a richer type
+func (c *ClansSummaries) GetClansSummaries(db models.DB, gameID string, publicIDs []string) ([]map[string]interface{}, error) {
+	resultMap := c.getCachedClansSummaries(gameID, publicIDs)
+	err := c.getAndCacheClansSummaries(db, gameID, resultMap)
+	if err != nil {
+		if _, ok := err.(*models.CouldNotFindAllClansError); !ok {
+			return nil, err
+		}
+	}
+	var result []map[string]interface{}
+	for _, publicID := range publicIDs {
+		if summary := resultMap[publicID]; summary != nil {
+			result = append(result, summary)
+		}
+	}
+	return result, err
+}
+
+func (c *ClansSummaries) getClanSummaryCacheKey(gameID, publicID string) string {
+	return fmt.Sprintf("%s/%s", gameID, publicID)
+}
+
+func (c *ClansSummaries) getClanSummaryCache(gameID, publicID string) map[string]interface{} {
+	clan, present := c.Cache.Get(c.getClanSummaryCacheKey(gameID, publicID))
+	if !present {
+		return nil
+	}
+	return clan.(map[string]interface{})
+}
+
+func (c *ClansSummaries) setClanSummaryCache(gameID, publicID string, clanPayload map[string]interface{}) {
+	ttl := c.TTL - c.TTLRandomError
+	if ttl < 0 {
+		ttl = 0
+	}
+	ttl += time.Duration(rand.Intn(int(2*c.TTLRandomError + 1)))
+	c.Cache.Set(c.getClanSummaryCacheKey(gameID, publicID), clanPayload, ttl)
+}
+
+// Return type map[string]map[string]interface{} maps public IDs to cached summaries.
```

Comment:
Missing function name at the start of the comment.

Eg:

`// getCachedClansSummaries reads clan summaries from the cache. It returns ...`

Reply:
Big function in next commit, this function doesn't exist anymore!